databricks/koalas/typedef/typehints.py

```diff
@@ -0,0 +1,235 @@
+# limitations under the License.
+#
+
+"""
+Utilities to deal with types. This is mostly focused on python3.
+"""
+import typing
+import datetime
+from inspect import getfullargspec
+
+import numpy as np
+import pandas as pd
+from pandas.api.types import is_datetime64_dtype, is_datetime64tz_dtype
+import pyarrow as pa
+import pyspark.sql.types as types
+from pyspark.sql.types import UserDefinedType
+
+try:
+    from pyspark.sql.types import to_arrow_type, from_arrow_type
+except ImportError:
+    from pyspark.sql.pandas.types import to_arrow_type, from_arrow_type
+
+from databricks import koalas as ks  # For running doctests and reference resolution in PyCharm.
+from databricks.koalas.typedef.string_typehints import resolve_string_type_hint
+
+
+# A column of data, with the data type.
+class SeriesType(object):
+    def __init__(self, tpe):
+        self.tpe = tpe  # type: types.DataType
+
+    def __repr__(self):
+        return "SeriesType[{}]".format(self.tpe)
+
+
+class DataFrameType(object):
+    def __init__(self, tpe):
+        # Seems we cannot specify field names. I currently gave some default names
+        # `c0, c1, ... cn`.
+        self.tpe = types.StructType(
+            [types.StructField("c%s" % i, tpe[i]) for i in range(len(tpe))]
+        )  # type: types.StructType
+
+    def __repr__(self):
+        return "DataFrameType[{}]".format(self.tpe)
+
+
+# The type is a scalar type that is furthermore understood by Spark.
+class ScalarType(object):
+    def __init__(self, tpe):
+        self.tpe = tpe  # type: types.DataType
+
+    def __repr__(self):
+        return "ScalarType[{}]".format(self.tpe)
+
+
+# The type is left unspecified or we do not know about this type.
+class UnknownType(object):
+    def __init__(self, tpe):
+        self.tpe = tpe
+
+    def __repr__(self):
+        return "UnknownType[{}]".format(self.tpe)
+
+
+def _to_stype(tpe) -> typing.Union[SeriesType, DataFrameType, ScalarType, UnknownType]:
+    if isinstance(tpe, str):
+        # This type hint can happen when given hints are string to avoid forward reference.
+        tpe = resolve_string_type_hint(tpe)
+    if hasattr(tpe, "__origin__") and tpe.__origin__ == ks.Series:
+        inner = as_spark_type(tpe.__args__[0])
+        return SeriesType(inner)
+    if hasattr(tpe, "__origin__") and tpe.__origin__ == ks.DataFrame:
+        tuple_type = tpe.__args__[0]
+        if hasattr(tuple_type, "__tuple_params__"):
+            # Python 3.5.0 to 3.5.2 has '__tuple_params__' instead.
+            # See https://github.com/python/cpython/blob/v3.5.2/Lib/typing.py
+            parameters = getattr(tuple_type, "__tuple_params__")
+        else:
+            parameters = getattr(tuple_type, "__args__")
+        return DataFrameType([as_spark_type(t) for t in parameters])
+    inner = as_spark_type(tpe)
+    if inner is None:
+        return UnknownType(tpe)
+    else:
+        return ScalarType(inner)
+
+
+def as_spark_type(tpe) -> types.DataType:
```

Comment:
except this one. I refactored this as if-else to make it more readable. the logic should be the same.